core/pipe_test.go

```diff
@@ -359,6 +361,44 @@ func TestDataSources(t *testing.T) {
 	})
 }
 
+func TestDataSourcesFailure(t *testing.T) {
+	Convey("Given a data source", t, func() {
+		ctx := NewContext(nil)
+		srcs := newDataSources(NTBox, "test_component")
+		r, s := newPipe("test", 1)
+		srcs.add("test_node", r)
+		Reset(func() {
+			s.close()
+		})
+		stopped := make(chan error, 1)
+		go func() {
+			stopped <- srcs.pour(ctx, WriterFunc(func(ctx *Context, t *Tuple) error {
+				return errors.New("error")
+			}), 4)
+		}()
+		srcs.state.Wait(TSRunning)
+		Reset(func() {
+			srcs.stop(ctx)
+		})
+		t := &Tuple{
+			InputName: "some_component",
+			Data: data.Map{
+				"v": data.Int(1),
+			},
+		}
+
+		Convey("When writing a tuple to it and the connected node returns an error", func() {
+			So(s.Write(ctx, t), ShouldBeNil)
+			srcs.stop(ctx)
+			So(<-stopped, ShouldBeNil)
+
+			Convey("Then numError should be increased", func() {
+				So(srcs.numErrors, ShouldEqual, 1)
```

Comment:
I think that it is better to confirm  `srcs.numErrors` equals 0 before written, and talk with nobu-k, the test suite has an implicit premise that  `srcs` is initialized with `srcs.numErrors = 0`